app/models/generic_object_definition.rb

```diff
@@ -7,6 +16,21 @@ class GenericObjectDefinition < ApplicationRecord
   has_many  :generic_objects
 
   def defined_attributes
-    @defined_attributes ||= properties[:attributes].stringify_keys
+    properties[:attributes]
+  end
+
+  def properties=(props)
+    props = props.symbolize_keys
+    if props.key?(:attributes)
+      props[:attributes] = props[:attributes].each_with_object({}) do |(name, type), hash|
+        raise ArgumentError, "#{type} is not a recognized type" unless TYPE_MAP.key?(type.to_sym)
+        hash[name.to_s] = type.to_sym
+      end
+    end
```

Comment:
@Fryguy I prefer to keep my implementation because it requires only one loop. It is not hard to read either. Your suggested code is easy to read and looks simplified but requires three loops.

I am not worried about the passed-in `props`. It is unmodified due to the first line `props = props.symbolized_keys`.

I have updated the PR with your other recommendations. Please review again. Thanks.